### DISCUSSION TITLE
why is allowCustomCrateRegistries self-hosted only?

### DISCUSSION THREAD

**Discussion by rbtcollins** [renovatebot/renovate#18872] [Request Help]
We have a private Cargo repository, conceptually the same as a private Maven repo: there is an encrypted token in a HostRule for the http://HOST/ that the registry is stored on.

Accessing the registry metadata could be done with either the sparse API or a shallow git clone : the credentials we have will do that correctly. As far as I can tell renovate itself isn't dependent on cargo - it parses and processes the data itself, so it should be possible to do that cheaply without having to muck around with the cargo [net] configuration etc - though if you do call out to cargo, environment variables control this and allow overriding things without state management.

If there is no fundamental reason, and it is just a code gap, then I'd be happy enough to put together a PR to enable this, since it would be super useful and avoid having a github action setup just to thunk across.

**Comment by rarkins** [comment]
I believe this was just a code gap, and a PR would be welcome. 

If I remember correctly, private crate registries were implemented using git, and naturally we wanted to cache the clone instead of cloning once per package, and then we worried about accidentally leaking caches between users. Result: make it self hosted only initially.

**Reply by rbtcollins** [reply]
If ensuring hygiene is important - is there some containment already in place for Renovate aaS ? (E.g. a chroot / cgroup around the processing of each repository?) If so, contained caching for the git repos should be very easy: use a strong hash of the url *and the credentials* to the git repo as a cache location. After each run `git push -f` the origin branch from the isolated run of cargo back to the cache location, seeding the cache for the next run. I could implement this rather than the stateless isolation I described below, but it does depend on containment - without that there is too much room for someone using e.g. npm or pip or something to cause arbitrary code to run and access the cached repos, disclosing package names and versions and download urls.

**Comment by rbtcollins** [comment]
Ok so a few notes so far.

Artifactory (the private registry we're using) does support the sparse index protocol, so thats good. 

However, it supports it at a *different* URL space than the git index.

e.g. if the git index is at 
`https://COMPANY.jfrog.io/artifactory/git/NAME.git` then the sparse index is at `https://COMPANY.jfrog.io/artifactory/api/cargo/NAME/index`. 

I think I need to add registryAliases support for cargo to do that? And then we need to resolve by direct HTTP call rather than looking up on disk, but that should be about it. 

I'd also like to add documentation to guide others on this path.

**Reply by rbtcollins** [reply]
However that will only make it possible to detect the presence of the change. The requirement to call 'cargo update' when there is a lockfile will cause this to fail today because of the missing authentication support.

I think the following should work

0. use registryAliases to detect if we need to do this code path
1. we create a temporary sandbox dir to isolate cargo.
2. We copy the top level structure from your existing CARGO_HOME to it: config.toml, credentials.toml, 
3. We create `registry/index/{github.com-1ecc6299db9ec823,index.crates.io-93a9e06d8945f2c0}` as symlinks to the original Cargo structure (these are the public crates.io). So new indices will be downloaded into this sandbox structure, but public ones are shared. The global lock will not be visible to other cargo instance, but the crates-io git index is concurrent writer safe (its git); and for a sparse crates.io index sparse index files are written in-place with truncate, so concurrent cargo runs might stomp on each other but it will self correct - (see Summaries::parse) - with the last cargo completing the final write out of the file and so the system will recover without intervention. 
4. We write the hostRules values we've received to a .netrc file in another temp dir - see below - because git uses https://curl.se/libcurl/c/CURLOPT_NETRC.html to enable HTTP authentication, we can project the git CLI authentication to disk, sandboxed to the one invocation of cargo, safely.
5. when running cargo update we export CARGO_HOME and HOME to the temp dirs.
6. delete the temp dirs when finished

the file structure would be something like

* cargo-temps/{tempdir}/cargo/config.toml
* cargo-temps/{tempdir}/cargo/credentials.toml
* cargo-temps/{tempdir}/registry/cache/github.com-1ecc6299db9ec823 -> ~/.cargo/registry/cache/github.com-1ecc6299db9ec823
* cargo-temps/{tempdir}/registry/cache/index.crates.io-93a9e06d8945f2c0 -> ~/.cargo/registry/index/index.crates.io-93a9e06d8945f2c0
* cargo-temps{tempdir}/home/.netrc

It might be good to facilitate easy cleanup to add a date in there:

* cargo-temps/yyyy-mm-dd/{tempdir}/cargo/config.toml
* etc

And then include a background task that looks for not-today directories and deletes them, Just In Case something interrupts renovate and a directory is leaked.

What do you think?

**Comment by rarkins** [comment]
The first step would be making sure Renovate can lookup the sparse index over HTTP, including populating the hostRules (but the hostRules should be the easy part). Can Renovate detect the custom registry during the cargo extract phrase, or would it need to be configured using `registryUrls`?

The second step would be making sure `cargo update` is passed the necessary credentials it needs, by converting them from hostRules to a format `cargo` understands.

Right now I'm not sure why there's a need for any sandbox concept, or are you worried about Cargo leaking the details somewhere outside the repo directory that's not cleaned up?

**Reply by rbtcollins** [reply]
Cargo writes a cache to ~/.cargo/registry/ with content from each registry, which is a potential information leak between organisations. It does this whether or not the registry is sparse or git based.

Secondly the content of ~/.netrc is security sensitive - it will allow direct access to the private registry and its state on disk. As Cargo itself doesn't yet support authentication, we have to project that to disk to enable cargo's git usage. Unlike something in <repo>/.cargo/config.toml its not natively isolated per repository.

We could put the sandbox in the repository but then the build/update process for a repository might flag it and error :/.

**Reply by rarkins** [reply]
Is that location configurable, e.g. via environment variable? Our preferred approach is:
- have package managers to save data to our internal concept `cacheDir` (e.g. by setting `PIP_CACHE_DIR`) to enable reuse within the same repo/run
- Delete that cache directory later

Even if it's hardcoded, it will be run within "sidecar" containers in our hosted app, and each container is destroyed after each time it's run. Same would apply to `~/.netrc` if you mean it's added during the artifacts update stage inside the sidecar container.

**Reply by rbtcollins** [reply]
`CARGO_HOME` which I mentioned above sets that for Cargo; its not very granular but if you have a cache dir already thats great. Is it coupled to the lifetime of the repo working tree (e.g do they share fate, co-located in some fashion)? If so then we can persist this easily, and the only remaining complexity will be performance optimisations like seeding from a centralised public cargo to save bandwidth on first operation.

I'll look up the cacheDir bits.

**Reply by rarkins** [reply]
Example to copy: https://github.com/renovatebot/renovate/blob/4abd0abea82a8880a91c2e3c4e1216ffbc8459c6/lib/modules/manager/pip_requirements/artifacts.ts#L78-L80

**Reply by rbtcollins** [reply]
It looks like privateCacheDir would be more suitable? Then just needs an async helper like ensureCacheDir for it?

**Reply by rarkins** [reply]
In the hosted app, the cache dir is cleared out after each repo anyway.

If you prefer it to be cleared after each command (meaning if you run multiple cargo artifact updating commands in the same repo, they don't share cache) then that should be the default behavior today.

So I'm not sure if a private cache dir helps in any way.

**Reply by rbtcollins** [reply]
Ok. Ideally the cache would live for the whole repo, so that multiple crates in a single workspace reuse the cache.

**Reply by rarkins** [reply]
This might be enough:

```
 extraEnv: { 
   CARGO_HOME: await ensureCacheDir('cargo'), 
 }, 
```

**Comment by rbtcollins** [comment]
What is `getDependencyURL` used for? It seems like cargo dependencies urls for dependency grouping are extracted from the packages, not from the datasource. Is it perhaps a fallback?

**Reply by rarkins** [reply]
`dependencyUrl` is optional and populated if there's a public HTTP url for the package on the registry.  e.g. https://npmjs.com/package/lodash which is different from the `sourceUrl` which is https://github.com/lodash/lodash and different from `homepage` which is `https://lodash.com`.

**Reply by rbtcollins** [reply]
so a separate question then, but does it really make sense to have a depedencyUrl for private registries?

**Reply by rarkins** [reply]
The only purpose of it is to provide a link to the URL for users. It's entirely optional, and you don't need it if you don't think anyone would want to click on it

**Comment by rbtcollins** [comment]
So, I think I have something approximating a branch that should work, but it has a caveat I need to check on :) .

Your servers will need to run nightly cargo with two feature flags - -Z registry-auth and -Z sparse-registry. 

I could:

- hard code the use of this, which will keep working until the features are stable and then start warning, but be a trivial fix.
- make it opt-in for users of Cargo via a new setting in the manager

What do you think?

**Reply by rarkins** [reply]
When you say "need to run", do you mean for the `cargo update` commands which are used to update the lock file?

**Comment by samrogerson** [comment]
These features are now stable in cargo.

I've recently started experimenting with `renovate`. We make heavy use rust + sparse registries on artifactory, so I ended up implementing this feature, locally, and have verified it on our real repos. Happy to contribute a PR, but realise @rbtcollins also has a branch (only just came across this discussion!)

**Reply by rbtcollins** [reply]
I am just crazy bandwidth constrained at work. I say put your patch up - great! If its missing things we need I can adjust my patch to match.

What we need is sparse + authenticated registry support.

**Reply by samrogerson** [reply]
Sounds good, will get something up  :)

**Reply by samrogerson** [reply]
https://github.com/renovatebot/renovate/pull/21187

**Comment by nated0g** [comment]
Reviving this thread, since we have a use case for this.  We are using a private cargo registry and it has necessitated us migrating off of Mend hosted renovate onto our own self hosted renovate.

Now that we have sparse registries and registry auth in Cargo, are there any fundamental decision left to make here, or does this just need someone to put in the work?

**Reply by jamietanna** [reply]
I think this is a case of "we need to do it" - as I understand, https://github.com/renovatebot/renovate/pull/21187 is now here, so there's some work to wire it in - if we can see that this isn't something that leads to performance issues when using it (and doesn't have any risks of i.e. remote code execution) it's something I'm willing to remove as a gated feature and instead have on-by-default

**Reply by lucab** [reply]
I had a look at the internals of the `allowCustomCrateRegistries` settings and I noticed that, for the `sparse` registries implemented in https://github.com/renovatebot/renovate/pull/21187, the current gating logic is both redundant and overly broad: it gates a `git` clone which is already conditional on `if (!isSparseRegistry)`, and on the other end it blocks operations on `sparse` registries which do not involve `git` clones to begin with.
I opened https://github.com/renovatebot/renovate/pull/45632 to re-scope this global Cargo guard to `git` registries only, leaving `sparse` ones unaffected.
